src/Microsoft.PowerShell.Security/security/CertificateProvider.cs

```diff
@@ -2633,42 +2725,132 @@ private bool IncludeArchivedCerts()
             return includeArchivedCerts;
         }
 
-        // If it's Win8 or above, filter matching for certain properties is done by
-        // the certificate enumeration filter at the API level. In that case,
-        // filter.Purpose will be 'None' and MatchesFilter will return 'True'.
-        private static bool MatchesFilter(X509Certificate2 cert,
-                                   CertificateFilterInfo filter)
+        private static bool MatchesFilter(X509Certificate2 cert, CertificateFilterInfo filter)
         {
-            //
-            // no filter means, match everything
-            //
-            if ((filter == null) ||
-                (filter.Purpose == CertificatePurpose.NotSpecified) ||
-                (filter.Purpose == CertificatePurpose.All))
+            // No filter means, match everything
+            if (filter == null)
             {
                 return true;
             }
 
+            if (filter.Expiring > DateTime.MinValue && !SecuritySupport.CertIsExpiringInDays(cert, filter.Expiring))
+            {
+                return false;
+            }
+
+            if (filter.DnsName != null && !CertContainsName(cert, filter.DnsName))
+            {
+                return false;
+            }
+
+            if (filter.Eku != null && !CertContainsEku(cert, filter.Eku))
+            {
+                return false;
+            }
+
+            if (filter.SSLServerAuthentication && !CertIsSSLServerAuthentication(cert))
+            {
+                return false;
+            }
+
             switch (filter.Purpose)
             {
                 case CertificatePurpose.CodeSigning:
-                    if (SecuritySupport.CertIsGoodForSigning(cert))
-                    {
-                        return true;
-                    }
+                    return SecuritySupport.CertIsGoodForSigning(cert);
+
+                case CertificatePurpose.DocumentEncryption:
+                    return SecuritySupport.CertIsGoodForEncryption(cert);
 
+                case CertificatePurpose.NotSpecified:
+                case CertificatePurpose.All:
+                    return true;
+
+                default:
                     break;
+            }
 
-                case CertificatePurpose.DocumentEncryption:
-                    if (SecuritySupport.CertIsGoodForEncryption(cert))
+            return false;
+        }
+
+        /// <summary>
+        /// Check if the specified certificate has the name in DNS name list.
+        /// </summary>
+        /// <param name="cert">Certificate object.</param>
+        /// <param name="pattern">Wildcard pattern for DNS name to search.</param>
+        /// <returns>True on success, false otherwise.</returns>
+        internal static bool CertContainsName(X509Certificate2 cert, WildcardPattern pattern)
```

Comment:
Have you verified this will work with punycode?

Reply:
Do you mean that Windows PowerShell support also punycode in dnsname filter? I did not find this in docs https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.security/about/about_certificate_provider?view=powershell-6

Reply:
When the name in the cert is in punycode.  This is a compliance requirement.  you deleted the code to do this.

Reply:
Clear about compliance. Not clear what code do you mean. 
Also docs say:
> DnsName <Microsoft.PowerShell.Commands.DnsNameRepresentation>
This parameter gets certificates that have the specified domain name or name pattern in the DNSNameList property of the certificate. The value of this parameter can either be "Unicode" or "ASCII". Punycode values are converted to Unicode. Wildcard characters (*) are permitted.

It is not clear how punycode converted to Unicode. It seems this did unpublic code (in P/Invoke) which we removed long ago.

Reply:
I looked original code which was removed and see only one difference - DNSName parameter was DnsNameRepresentation type, now string.
In both cases DnsNameRepresentation is initialized by one constructor with string parameter which assigned to DNSname and punycode - no conversion. So new code works for punycode.
Question is should we change type from string to DnsNameRepresentation?